# **Assignment 4**: TextBlob: Sentences and Semantics
Emily Daskas

In [ ]:
import pandas as pd
from textblob import TextBlob
import nltk
nltk.download('punkt')
nltk.download('brown')
nltk.download('stopwords')
from nltk.corpus import stopwords
import spacy 
nlp = spacy.load('en')
from textblob import Word
import nltk
nltk.download('wordnet')
from datascience import *
from pathlib import Path
!pip install textatistic
from textatistic import Textatistic

#ReadtheRomeo & Juliettext.
#Retrieveonlythefirst1,000charactersfromthetext
textTotal = open('/RomeoJuliet.txt').read() 
blobTotal = TextBlob(textTotal)
######################################
numChars= 1000
text1000 = textTotal[0:numChars+1]
blob1000 =TextBlob(text1000)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Package brown is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


#### Analysis of Sentences 

In [ ]:
print("1) Total number of sentences: ", len(blobTotal.sentences)) 

1) Total number of sentences:  3169


In [ ]:
stringQueen = "Queen"
count = 0
for sentence in blobTotal.sentences:
  if stringQueen in sentence:
    count += 1
print('2) Total number of sentences that contain the word "Queen":', count)

2) Total number of sentences that contain the word "Queen": 1


In [ ]:
sentence_lens = []

for sentence in blobTotal.sentences:
  sentence_lens.append(len(sentence))

t1 = Table().with_column("Sentence", blobTotal.sentences, "Length", sentence_lens)
sorted_longest_length_table = t1.sort("Length", descending = True)

print("3) 10 longest sentences in Romeo and Juliet:")
sorted_longest_length_table.show(10)


3) 10 longest sentences in Romeo and Juliet:


/usr/local/lib/python3.7/dist-packages/datascience/tables.py:483: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  values = np.array(tuple(values))


In [ ]:
sorted_shortest_length_table = t1.sort("Length", descending = False)

print("4) 10 shortest sentences in Romeo and Juliet:")
sorted_shortest_length_table.show(10)

4) 10 shortest sentences in Romeo and Juliet:


#### Analysis of Sentiments

In [ ]:
pos_polarity = 0
neg_polarity = 0
neutral_polarity = 0
total_sentences = 0
for sentence in blobTotal.sentences:
  if sentence.polarity > 0:
    pos_polarity += 1
  elif sentence.polarity == 0:
    neutral_polarity += 1
  else:
    neg_polarity += 1

  total_sentences += 1

print("1)")
print("Total number of Positive sentences:", pos_polarity)
print("Total number of Negative sentences:", neg_polarity)
print("Total number of Neutral sentences:", neutral_polarity)
print("Total number of sentences:", total_sentences, "is the same as sum of positive, negative, and neutral polarity sentences:", pos_polarity + neg_polarity + neutral_polarity)

1)
Total number of Positive sentences: 666
Total number of Negative sentences: 279
Total number of Neutral sentences: 2224
Total number of sentences: 3169 is the same as sum of positive, negative, and neutral polarity sentences: 3169


In [ ]:
sentence_polarities = []

for sentence in blobTotal.sentences:
  sentence_polarities.append(sentence.polarity)

polarity_table = Table().with_column("Sentence", blobTotal.sentences, "Polarity", sentence_polarities)

/usr/local/lib/python3.7/dist-packages/datascience/tables.py:483: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  values = np.array(tuple(values))


In [ ]:
sorted_high_polarity_table = polarity_table.sort("Polarity", descending = True)

print("2) 5 highest polarities")
sorted_high_polarity_table.show(5)

2) 5 highest polarities


In [ ]:
sorted_low_polarity_table = polarity_table.sort("Polarity", descending = False)

print("3) 5 lowest polarities")
sorted_low_polarity_table.show(5)

3) 5 lowest polarities


In [ ]:
print("4)")

for i in range(0,2):
  print("English:", sorted_longest_length_table['Sentence'][i])
  print("Japanese:", sorted_longest_length_table['Sentence'][i].translate(to='ja'))
  print()

4)
English: And in this state she gallops night by night
Through lovers’ brains, and then they dream of love;
O’er courtiers’ knees, that dream on curtsies straight;
O’er lawyers’ fingers, who straight dream on fees;
O’er ladies’ lips, who straight on kisses dream,
Which oft the angry Mab with blisters plagues,
Because their breaths with sweetmeats tainted are:
Sometime she gallops o’er a courtier’s nose,
And then dreams he of smelling out a suit;
And sometime comes she with a tithe-pig’s tail,
Tickling a parson’s nose as a lies asleep,
Then dreams he of another benefice:
Sometime she driveth o’er a soldier’s neck,
And then dreams he of cutting foreign throats,
Of breaches, ambuscados, Spanish blades,
Of healths five fathom deep; and then anon
Drums in his ear, at which he starts and wakes;
And, being thus frighted, swears a prayer or two,
And sleeps again.
Japanese: そしてこの状態で彼女は夜ごとにギャロップします
恋人たちの頭脳を通して、そして彼らは愛を夢見ます。
カーテシーをまっすぐに夢見る、廷臣の膝。
手数料をまっすぐに夢見ている他の弁護士の指。
キスを真っ直ぐに夢見る女性の唇、
水ぶくれに悩まされ

In [ ]:
ngrams_polarities = []
all_ngrams = blobTotal.ngrams(n = 5)
for ngram in all_ngrams:
  ngram_str = str(ngram)
  ngram_blob = TextBlob(ngram_str)
  ngrams_polarities.append(ngram_blob.polarity)

In [ ]:
print("5) Top 10 nGrams (n=5) with highest polarity")
ngram_polarity_table = Table().with_column("nGrams", all_ngrams, "Polarity", ngrams_polarities)
sorted_ngram_polarity_table = ngram_polarity_table.sort("Polarity", descending = True)
sorted_ngram_polarity_table.show(10)

5) Top 10 nGrams (n=5) with highest polarity


### Stemming, Lemmatization, Synonyms, Antonyms

In [ ]:
anticipation = Word('anticipation')
angry = Word('angry')

In [ ]:
print("1)", anticipation.stem())

print("2)", anticipation.lemmatize())

synonyms = set()
for synset in angry.synsets:
  for lemma in synset.lemmas():
    synonyms.add(lemma.name())
print("3)",synonyms)

lemmas = angry.synsets[0].lemmas()
print("4)", lemmas[0].antonyms())

1) anticip
2) anticipation
3) {'angry', 'raging', 'tempestuous', 'wild', 'furious'}
4) [Lemma('unangry.a.01.unangry')]


In [ ]:
romeo = nlp(Path('/RomeoJuliet.txt').read_text())
hamlet = nlp(Path('/Hamlet.txt').read_text()) 
les_mis = nlp(Path('/Les Miserable.txt').read_text()) 


/usr/lib/python3.7/runpy.py:193: ModelsWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if available.
  "__main__", mod_spec)


0.9847134017715589

In [ ]:
print("6)")
print("Romeo and Juliet vs Hamlet:", romeo.similarity(hamlet))
print("Romeo and Juliet vs Les Mis:", romeo.similarity(les_mis))
print("Hamlet vs Les Mis:", hamlet.similarity(les_mis))

6)
Romeo and Juliet vs Hamlet: 0.9847134017715589
Romeo and Juliet vs Les Mis: 0.9114393770895828
Hamlet vs Les Mis: 0.937290224741481


/usr/lib/python3.7/runpy.py:193: ModelsWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if available.
  "__main__", mod_spec)
/usr/lib/python3.7/runpy.py:193: ModelsWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if

The two shakespear texts were similar with 98.
When WS was compared against VH, the scores were lower with 91 and 93.

### Readibility Assessment

In [ ]:
text_total_romeo = open('/RomeoJuliet.txt').read() 
text_total_hamlet = open('/Hamlet.txt').read() 
text_total_les_mis = open('/Les Miserable.txt').read() 
all_texts = [text_total_romeo, text_total_hamlet, text_total_les_mis]
text_names = ["Romeo and Juliet", "Hamlet", "Les Mis"]

In [ ]:
idx = 0
for text in all_texts:
  print("Readability of ", text_names[idx])
  print(Textatistic(text).dict())
  idx+=1 

Readability of  Romeo and Juliet
{'char_count': 131781, 'word_count': 29318, 'sent_count': 3403, 'sybl_count': 35399, 'notdalechall_count': 6922, 'polysyblword_count': 1042, 'flesch_score': 95.94310121673479, 'fleschkincaid_score': 2.0174821979546937, 'gunningfog_score': 4.867787989863079, 'smog_score': 6.29026977403745, 'dalechall_score': 7.791850713812807}
Readability of  Hamlet
{'char_count': 160543, 'word_count': 35310, 'sent_count': 4098, 'sybl_count': 43997, 'notdalechall_count': 8021, 'polysyblword_count': 1459, 'flesch_score': 92.67598280115092, 'fleschkincaid_score': 2.473442610171446, 'gunningfog_score': 5.099348875241379, 'smog_score': 6.537782921689099, 'dalechall_score': 7.6507209880858635}
Readability of  Les Mis
{'char_count': 562037, 'word_count': 122869, 'sent_count': 7965, 'sybl_count': 157191, 'notdalechall_count': 25994, 'polysyblword_count': 6652, 'flesch_score': 82.94548596276992, 'fleschkincaid_score': 5.522374808844765, 'gunningfog_score': 8.33600414022729, 'smo

In terms of years of education required to understand the text, Les Mis would need to most years: 8.35, followed by Hamlet and then Romeo and Juliet, both requiring around 6 years.

Again, according to flesh reading ease score, Les Mis is the least readable, as it is the only one < 90, with a score of 82.9, indicatingit is not readable by 5th graders. 

In most all categories, Les Mis is the least readable.